tango/step.py

```diff
@@ -359,15 +352,250 @@ def unique_id(self) -> str:
+            elif isinstance(o, str):
+                return  # Confusingly, str is an Iterable of itself, resulting in infinite recursion.
+            elif isinstance(o, dict):
+                yield from dependencies_internal(o.values())
+            elif isinstance(o, Iterable):
+                yield from itertools.chain(*(dependencies_internal(i) for i in o))
+            else:
+                return
+
+        return dependencies_internal(self.kwargs.values())
+
+    @property
+    def dependencies(self) -> Set["Step"]:
+        """
+        Returns a set of steps that this step depends on. This does not return recursive dependencies.
+        """
+        return set(self._ordered_dependencies())
+
+    @property
+    def recursive_dependencies(self) -> Set["Step"]:
+        """
+        Returns a set of steps that this step depends on. This returns recursive dependencies.
+        """
+        seen = set()
+        steps = list(self.dependencies)
+        while len(steps) > 0:
+            step = steps.pop()
+            if step in seen:
+                continue
+            seen.add(step)
+            steps.extend(step.dependencies)
+        return seen
+
+
+class WithUnresolvedSteps(CustomDetHash):
+    """
+    This is a helper class for some scenarios where steps depend on other steps.
+
+    Let's say we have two steps, :class:`ConsumeDataStep` and :class:`ProduceDataStep`. The easiest way to make
+    :class:`ConsumeDataStep` depend on :class:`ProduceDataStep` is to specify ``Produce`` as one of the arguments
+    to the step. This works when ``Consume`` takes the output of ``Produce`` directly, or if it takes
+    it inside standard Python container, like a list, set, or dictionary.
+
+    But what if the output of :class:`ConsumeDataStep` needs to be added to a complex, custom data
+    structure? :class:`WithUnresolvedSteps` takes care of this scenario.
+
+    For example, this works without any help:
+
+    .. code-block:: Python
+
+        class ProduceDataStep(Step[MyDataClass]):
+            def run(self, ...) -> MyDataClass
+                ...
+                return MyDataClass(...)
+
+        class ConsumeDataStep(Step):
+            def run(self, input_data: MyDataClass):
+                ...
+
+        produce = ProduceDataStep()
+        consume = ConsumeDataStep(input_data = produce)
+
+    This scenario needs help:
+
+    .. code-block:: Python
+
+        @dataclass
+        class DataWithTimestamp:
+            data: MyDataClass
+            timestamp: float
+
+        class ProduceDataStep(Step[MyDataClass]):
+            def run(self, ...) -> MyDataClass
+                ...
+                return MyDataClass(...)
+
+        class ConsumeDataStep(Step):
+            def run(self, input_data: DataWithTimestamp):
+                ...
+
+        produce = ProduceDataStep()
+        consume = ConsumeDataStep(
+            input_data = DataWithTimestamp(produce, time.now()))
```

Comment:
```suggestion
        consume = ConsumeDataStep(
            input_data = DataWithTimestamp(produce, time.now())
        )
```

Reply:
I don't know why I can't commit these suggestions from the GitHub UI.